File: packages/cortex-dart/lib/src/resources/pools.dart
/// Pools management resource.
library;

import '../errors.dart';
import '../http_client.dart';
import '../types.dart';

/// Provides access to pool management endpoints.
class PoolsResource {
  final CortexHttpClient _client;
  final String _baseUrl;

  /// Creates a [PoolsResource].
  PoolsResource({
    required CortexHttpClient client,
    required String baseUrl,
  })  : _client = client,
        _baseUrl = baseUrl;

  /// Lists all pools.
  Future<List<Pool>> list() async {
    final json = await _client.get('$_baseUrl/admin/pools');
    final data =
        json['data'] as List<dynamic>? ?? json['pools'] as List<dynamic>?;
    if (data != null) {
      return data
          .map((e) => Pool.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return <Pool>[];
  }

  /// Creates a new pool.
  ///
  /// [name] is the pool name.
  /// [description] is an optional description.
  Future<Pool> create({
    required String name,
    String? description,
  }) async {
    if (name.isEmpty) {
      throw const CortexValidationException(
        message: 'Name must not be empty.',
        parameter: 'name',
      );
    }

    final body = <String, dynamic>{'name': name};
    if (description != null) body['description'] = description;

    final json = await _client.post('$_baseUrl/admin/pools', body: body);
    return Pool.fromJson(json);
  }

  /// Updates a pool.
  ///
  /// [id] is the pool identifier.
  /// [name] is the new name (optional).
  /// [description] is the new description (optional).
  Future<Pool> update({
    required String id,
    String? name,
    String? description,
  }) async {
    if (id.isEmpty) {
      throw const CortexValidationException(
        message: 'ID must not be empty.',
        parameter: 'id',
      );
    }

    final body = <String, dynamic>{};
    if (name != null) body['name'] = name;
    if (description != null) body['description'] = description;

    final json =
        await _client.patch('$_baseUrl/admin/pools/$id', body: body);
    return Pool.fromJson(json);
  }

  /// Deletes a pool.
  Future<void> delete(String id) async {
    if (id.isEmpty) {
      throw const CortexValidationException(
        message: 'ID must not be empty.',
        parameter: 'id',
      );
    }
    await _client.delete('$_baseUrl/admin/pools/$id');
  }

  /// Adds a backend to a pool.
  ///
  /// [poolId] is the pool identifier.
  /// [backendId] is the backend identifier to add.
  Future<Map<String, dynamic>> addBackend({
    required String poolId,
    required String backendId,
  }) async {
    if (poolId.isEmpty) {
      throw const CortexValidationException(
        message: 'Pool ID must not be empty.',
        parameter: 'poolId',
      );
    }
    if (backendId.isEmpty) {
      throw const CortexValidationException(
        message: 'Backend ID must not be empty.',
        parameter: 'backendId',
      );
    }

    return _client.post(
      '$_baseUrl/admin/pools/$poolId/backends',
      body: {'backendId': backendId},
    );
  }

  /// Removes a backend from a pool.
  Future<void> removeBackend({
    required String poolId,
    required String backendId,
  }) async {
    if (poolId.isEmpty) {
      throw const CortexValidationException(
        message: 'Pool ID must not be empty.',
        parameter: 'poolId',
      );
    }
    if (backendId.isEmpty) {
      throw const CortexValidationException(
        message: 'Backend ID must not be empty.',
        parameter: 'backendId',
      );
    }
    await _client.delete(
      '$_baseUrl/admin/pools/$poolId/backends/$backendId',
    );
  }
}

File: packages/cortex-js/src/resources/pools.ts
import type { CortexClient } from '../client.js';
import type {
  Pool,
  PoolCreateParams,
  PoolUpdateParams,
  PoolListResponse,
  PoolBackend,
  PoolAddBackendParams,
  DeleteResponse,
  RequestOptions,
} from '../types.js';
import { ValidationError } from '../errors.js';

export class Pools {
  private client: CortexClient;

  constructor(client: CortexClient) {
    this.client = client;
  }

  /**
   * List all pools.
   */
  async list(options?: RequestOptions): Promise<PoolListResponse> {
    const url = `${this.client._adminBaseUrl}/admin/pools`;
    return this.client._request<PoolListResponse>('GET', url, {
      requestOptions: options,
    });
  }

  /**
   * Create a new pool.
   */
  async create(params: PoolCreateParams, options?: RequestOptions): Promise<Pool> {
    if (!params.name || typeof params.name !== 'string') {
      throw new ValidationError('name is required and must be a string', 'name');
    }

    const url = `${this.client._adminBaseUrl}/admin/pools`;
    return this.client._request<Pool>('POST', url, {
      body: params,
      requestOptions: options,
    });
  }

  /**
   * Update a pool by ID.
   */
  async update(id: string, params: PoolUpdateParams, options?: RequestOptions): Promise<Pool> {
    if (!id) throw new ValidationError('id is required', 'id');

    const url = `${this.client._adminBaseUrl}/admin/pools/${encodeURIComponent(id)}`;
    return this.client._request<Pool>('PATCH', url, {
      body: params,
      requestOptions: options,
    });
  }

  /**
   * Delete a pool by ID.
   */
  async delete(id: string, options?: RequestOptions): Promise<DeleteResponse> {
    if (!id) throw new ValidationError('id is required', 'id');

    const url = `${this.client._adminBaseUrl}/admin/pools/${encodeURIComponent(id)}`;
    return this.client._request<DeleteResponse>('DELETE', url, {
      requestOptions: options,
    });
  }

  /**
   * Add a backend to a pool.
   */
  async addBackend(
    poolId: string,
    params: PoolAddBackendParams,
    options?: RequestOptions,
  ): Promise<PoolBackend> {
    if (!poolId) throw new ValidationError('poolId is required', 'poolId');
    if (!params.backendId) throw new ValidationError('backendId is required', 'backendId');

    const url = `${this.client._adminBaseUrl}/admin/pools/${encodeURIComponent(poolId)}/backends`;
    return this.client._request<PoolBackend>('POST', url, {
      body: params,
      requestOptions: options,
    });
  }

  /**
   * Remove a backend from a pool.
   */
  async removeBackend(
    poolId: string,
    backendId: string,
    options?: RequestOptions,
  ): Promise<DeleteResponse> {
    if (!poolId) throw new ValidationError('poolId is required', 'poolId');
    if (!backendId) throw new ValidationError('backendId is required', 'backendId');

    const url = `${this.client._adminBaseUrl}/admin/pools/${encodeURIComponent(poolId)}/backends/${encodeURIComponent(backendId)}`;
    return this.client._request<DeleteResponse>('DELETE', url, {
      requestOptions: options,
    });
  }
}
